### ISSUE TITLE
a problem with creating a schema with drf 3.10.3

### ISSUE THREAD

**Issue by fannyhub** [tfranzel/drf-spectacular#796]
**Describe the bug**
I have used drf version 3.10.3 as suggested and my previous problem has been resolved.
Alas, the schema still does not get built, failing with:
`AssertionError: Incompatible AutoSchema used on View <class 'view-name'>. Is DRF's DEFAULT_SCHEMA_CLASS pointing to "drf_spectacular.openapi.AutoSchema" or any other drf-spectacular compatible AutoSchema?`

This happens although I have set DEFAULT_SCHEMA_CLASS in REST_FRAMEWORK settings to drf_spectacular.openapi.AutoSchema.

**To Reproduce**
create a project with drf 3.10.3 (Django 2.2, Python 3.8/3.9)
Execute ./manage.py spectacular --file schema.yml

**Expected behavior**
I would expect the schema to pass smoothly 

**The same applies to 3.11.2 drf version**

**Comment by tfranzel** [comment]
Hi,

every time this comes up, it is a different `settings.py` file used then what was assumed or a `settings.py` file that overrides what you originally had. This is usually due to environment variable `DJANGO_SETTINGS_MODULE`.

```python
REST_FRAMEWORK = {
    # YOUR SETTINGS
    'DEFAULT_SCHEMA_CLASS': 'drf_spectacular.openapi.AutoSchema',
}
```
 will 100% work. Just make sure you have the right settings file loaded.

**Comment by fannyhub** [comment]
thank you for such a quick reply. The variable seems not to be overriden:
![image](https://user-images.githubusercontent.com/5873729/186880197-5c662d64-318b-409d-9704-80a43a99c750.png)

**Comment by tfranzel** [comment]
That is the usual fallacy. running `./manage.py runserver` and `./manage.py shell` might load different settings depending on your setup. Sometimes simply by the prescence of a different `DJANGO_SETTINGS_MODULE`  environment variable. It is a common issue. Try to use the debugger on your actual webserver process if possible. I'm quite confident it will be different in there.

**Comment by fannyhub** [comment]
Alas, it is not possible that we touch this variable. btw, we do not use runserver

**Comment by tfranzel** [comment]
with `runserver` I meant your server process. Otherwise just try to `print` that variable to console in the process, where this happens. 

Either you have some custom schema class on views OR your `DEFAULT_SCHEMA_CLASS` is not properly set in the environment. I cannot see any other explanation. Again, this has come up at least a dozen times. Every time it was a misconfigured `settings.py`. This works for hundreds of people without issue so it has to be something in your setup/environment I suppose.

**Comment by fannyhub** [comment]
Thank you very much for another reply. By not touching I meant that we didn't change the DEFAULT_SCHEMA_CLASS - it was introduced very recently, to see if transitioning to drf-spectacular was possible in our project.
 The thing is that the problem disappears with the upgrade of rest framework to version 3.12.4 or 3.13 and schema gets generated. So I don't believe it depends upon our code. Please consider that some other users have encountered it, too and solved it by drf's upgrade: https://stackoverflow.com/questions/64545861/drf-spectacular-is-using-the-wrong-autoschema-to-generate-swagger . Furthermore, as the AssertionError is raised by the line (in drf-spectacular) ` assert isinstance(view.schema, AutoSchema)` I found out that the view.schema is drf's ManualSchema. in version 3.10.3, ManualSchema usage in ObtainTokenView looks like that:
![image](https://user-images.githubusercontent.com/5873729/187311425-161ef75f-c05a-4525-ba3f-7e59452f467f.png)

meanwhile in version 3.13.1 version of drf, this code looks as follows:
![image](https://user-images.githubusercontent.com/5873729/187313736-be3fba4d-4545-4480-91d5-fc3f4bed94d7.png)

It seems that: 
1) in the first case - i.e. drf 3.10.3. -  the condition (if coreapi is not None and coreschema is not None) is True
2)  while in the second case - i.e. drf 3.13.1 -  (if coreapi.is_enabled() - the condition is False. In is_enabled() it is checked if api_settings.DEFAULT_SCHEMA_CLASS ,  is subclass of rest_framework.schemas.coreapi.AutoSchema. Debugging shows that api_settings.DEFAULT_SCHEMA_CLASS is drf_spectacular.openapi.AutoSchema.

**Comment by tfranzel** [comment]
> https://stackoverflow.com/questions/64545861/drf-spectacular-is-using-the-wrong-autoschema-to-generate-swagger

Thanks @fannyhub for digging in. Sorry that I nudged you in the wrong direction repeatedly. 99% of raised&resolved issues were indeed wrong configs. I was not aware of that SO question and apparently nobody raised it here. 

Your analysis is absolutely correct. It is indeed a bug in DRF < 3.12. I added an extension that fixes that bug for prior versions. It will be in the next release (`0.24.0`). Until then, you can easily fix it yourself by emulating what the extension will do:

```python

        from drf_spectacular.openapi import AutoSchema
        from rest_framework.authtoken.views import ObtainAuthToken

        class FixedObtainAuthToken(ObtainAuthToken):
            schema = AutoSchema()
```

and using that in your `urlpatterns` instead of `ObtainAuthToken/obtain_auth_token`

**Comment by fannyhub** [comment]
Thank you, no more problems with creating schema :)

**Comment by kelvinmacharia254** [comment]
AssertionError: Incompatible AutoSchema used on View <class 'Vendors.views.PurchaseOrderListCreateView'>. Is DRF's DEFAULT_SCHEMA_CLASS pointing to "drf_spectacular.openapi.AutoSchema" or any other drf-spectacular compatible AutoSchema?

Still have this issue in Django reset  3.14.0 and drf_spectacular 0.27.0 . 

I have have configured everything right as discussed here and as per documentation but no solution:

fork repo @:

https://github.com/Taruna21/Vendor_Management_System

**Comment by tfranzel** [comment]
this is an uncommon error and the message is exactly right. Some view, even from a library may set a custom schema class on a view and that provokes the error. If you are not doing it, some library does probably. This happened once before but I don't recall which lib. Maybe updating your libs will already fix the issue.

**Comment by kelvinmacharia254** [comment]
This is a recent project and has all the latest packages in my
dependencies. Please run my code an see

*Kind Regards,*
*Kelvin Macharia*

**Comment by tfranzel** [comment]
please find a minimal reproduction (the offending view) including stacktrace and create a new issue. I have very limited time to look at this.

**Comment by raparkhurst** [comment]
I am having this error with Django Rest Framework 3.15.2, Django 5.1.6 and Python 3.13.1

I have verified that I am setting:
REST_FRAMEWORK = {
    'DEFAULT_SCHEMA_CLASS': 'drf_spectacular.openapi.AutoSchema',
}

Only once in my settings.py file and I am currently not using environment variables (I moved away from it to help rule that out).

I have also verified that when calling:
print(REST_FRAMEWORK.get('DEFAULT_SCHEMA_CLASS'))

in settings.py it shows the right value("drf_spectacular.openapi.AutoSchema")

**Comment by kelvinmacharia254** [comment]
Hi Robert,

I hope my response is not too late.

remove the import *'from rest_framework.schemas import openapi'*  from '
*settings.py'* if it is there. It may be the one causing conflict in
setting drf schema class *'DEFAULT_SCHEMA_CLASS':
'drf_spectacular.openapi.AutoSchema'*

*Kind Regards,*
*Kelvin Macharia*

**Comment by JAtoms** [comment]
I am also facing this issue. Below is my REST_FRAMEWORK and I do not have 
`" import *'from rest_framework.schemas import openapi'*`  in `*settings.py'*`


```
REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework_simplejwt.authentication.JWTAuthentication',
    ),
    'DEFAULT_PERMISSION_CLASSES': (
        'rest_framework.permissions.IsAuthenticated',
    ),
    'DEFAULT_RENDERER_CLASSES': (
        'rest_framework.renderers.JSONRenderer',
        'rest_framework.renderers.BrowsableAPIRenderer',
    ),
    'DEFAULT_SCHEMA_CLASS': 'drf_spectacular.openapi.AutoSchema',
}
```

**Comment by tfranzel** [comment]
> I am having this error with Django Rest Framework 3.15.2, Django 5.1.6 and Python 3.13.1

almost exactly that combination is part of the test suite, so it does not look like a general issue. 

@JAtoms @raparkhurst, its hard do help with so little information. What does the assertion say? Which view? Please show the view and the assert message. 

This issue was about a broken auth view that hardcoded the default schema in DRF<3.12.0 and that has long been fixed. 

Either there is some weird thing going on with your settings, or you might use a library that also has a hardcoded default schema one some view (which is more unlikely). However, we have tests for the most used libs.

Is only one view offending or all of them? Are you setting `schema` on any view manually? This message only comes up, when spectacular finds a foreign schema base class, that is unknown to us. Either the `settings.py` are not applied properly, overridden, or some view sets the `schema` class variable to something unexpected. The assert message should really point in the right direction.

**Comment by raparkhurst** [comment]
Sorry I didn't reply before now.

In settings.py I do not have that line ()

I have it setup right now that in my site urls.py file I have the following relevant lines:
`from drf_spectacular.views import SpectacularAPIView, SpectacularRedocView, SpectacularSwaggerView`
` `
`urlpatterns = [
    path('api/schema/', SpectacularAPIView.as_view(), name='schema'),
    path('api/schema/swagger-ui/', SpectacularSwaggerView.as_view(url_name='schema'), name='swagger-ui'),
    path('api/schema/redoc/', SpectacularRedocView.as_view(url_name='schema'), name='redoc'),
]`

I can provide the full contents if you need (there isn't too much more than this).

The assertion in the logs when I hit `127.0.0.1:8000/api/schema/swagger-ui` looks to be:
AssertionError: Incompatible AutoSchema used on View <class 'drf_spectacular.views.SpectacularAPIView'>. Is DRF's DEFAULT_SCHEMA_CLASS pointing to "drf_spectacular.openapi.AutoSchema" or any other drf-spectacular compatible AutoSchema?

There is also this line which seems interesting:
2025-02-27 03:55:44 [DEBUG   ] Exception while resolving variable 'exception_notes' in template 'unknown'.

The libraries in my `requirements.txt` file are:
`asgiref==3.8.1
attrs==25.1.0
bcrypt==4.2.1
boto3==1.36.17
botocore==1.36.17
certifi==2025.1.31
cffi==1.17.1
charset-normalizer==3.4.1
configobj==5.0.9
coreapi==2.3.3
coreschema==0.0.4
cryptography==44.0.1
decorator==5.1.1
Deprecated==1.2.18
dj-rest-auth==7.0.1
Django==5.1.6
django-allauth==65.4.1
django-cors-headers==4.7.0
django-crontab==0.7.1
django-simple-history==3.8.0
djangorestframework==3.15.2
djangorestframework_simplejwt==5.4.0
drf-spectacular==0.27.2
drf-yasg==1.21.8
environ==1.0
fabric==3.2.2
idna==3.10
inflection==0.5.1
invoke==2.2.0
itypes==1.2.0
Jinja2==3.1.5
jmespath==1.0.1
jsonschema==4.23.0
jsonschema-specifications==2024.10.1
MarkupSafe==3.0.2
openapi-codec==1.3.2
packaging==24.2
paramiko==3.5.1
psycopg2-binary==2.9.10
pycgi==0.1a3
pycparser==2.22
Pygments==2.19.1
PyJWT==2.10.1
PyNaCl==1.5.0
python-dateutil==2.9.0.post0
python-dotenv==1.0.1
pytz==2025.1
PyYAML==6.0.2
referencing==0.36.2
requests==2.32.3
rpds-py==0.22.3
s3transfer==0.11.2
setuptools==75.8.0
simplejson==3.19.3
six==1.17.0
sqlparse==0.5.3
uritemplate==4.1.1
urllib3==2.3.0
watchtower==3.3.1
wrapt==1.17.2
`

**Comment by tfranzel** [comment]
@raparkhurst the fact that this fails for `drf_spectacular.views.SpectacularAPIView` tells me there is something fundamentally wrong in the setup. This is one of our views and has to work. If this fails, likely all views would fail.

I can see you have `drf-yasg` in there. It cannot work with both libraries active. You need to comment out everything related to `drf-yasg` including in the config. I have the feeling some yasg configuration is breaking everything here.

On a clean setup, it is almost impossible to break `SpectacularAPIView`. I'm pretty sure yasg injects its own schema class implementation somewhere and that breaks the whole thing.

**Comment by raparkhurst** [comment]
Ah, I just set up a test project and it's working.  OK sorry about that,
I'll start going through my project and see where I messed up (starting
with getting rid of drf-yasg).

**Comment by JAtoms** [comment]
Resolved, I had this `from drf_spectacular.openapi import AutoSchema` in my import in settings.py. Removed it and everything works fine

**Comment by kelvinmacharia254** [comment]
I am glad my tip helped
